test/distributed/test_c10d_nccl.py

```diff
@@ -1243,6 +1243,19 @@ def test_close_pg(self):
 
         del pg
 
+    @requires_nccl()
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
+    def test_nan_assert(self):
+        os.environ["TORCH_NCCL_NAN_CHECK"] = "1"
+
+        store = c10d.FileStore(self.file_name, self.world_size)
+        pg = self._create_process_group_nccl(store, self.opts())
+        device = self.rank_to_GPU[self.rank][0]
+        nan_tensor = torch.full((2, 3), float("nan"), device=device)
```

Comment:
Let's beef up the test to cover all the dtypes we support in c10d comms, and also cover a range of sizes of tensors, and put the nan in just one place in the tensor (a random place).